include/glow/Optimizer/GraphOptimizer/GraphOptimizer.h

```diff
@@ -57,18 +76,24 @@ void convertPlaceholdersToConstants(Function *F,
 /// in context \p bindings.
 void profileQuantization(PlaceholderBindings &bindings, Function *F);
 
-/// Optimize the Function \p F given compilation options \p cctx for Backend \B.
-/// \returns success if all nodes in the final resulting optimized Function are
-/// supported by \p B; if not, this represents a compiler error.
+/// Optimize the Function \p F given compilation options \p cctx and pipeline
+/// \p pipeline for Backend \B. \returns success if all nodes in the final
+/// resulting optimized Function are supported by \p B; if not, this represents
+/// a compiler error.
 llvm::Error optimizeFunction(Function *F, const Backend &B,
-                             CompilationContext &cctx);
+                             CompilationContext &cctx,
```

Comment:
I am not sure if an error should be returned. If an optimization is supported, then apply the optimization. If the optimization is not supported, can we just reports some warning and skip it? IMO, optimization is kind of "optional". Unless some opts are forced due to some execution correctness reason, we don't have to report a compiler error ?

Reply:
FYI returning an Error pre-dates this PR. I agree optimization passes are somewhat optional. Though one exception is lowering, since a backend might not support an op unless it gets lowered.

Either way, the motivation for returning an Error is essentially for reporting a compiler bug. For example, we return an error if the `cctx` is not set up correctly for a specific compilation setup (e.g. if `convertToFP16` is set when in profiling mode). We also return an error if after all of the optimization passes the backend reports it does not support one of the Nodes.

We could try to pull some of this error checking out into the callers if we decide it's problematic to return an Error here.